packages/opentelemetry-basic-tracer/src/Span.ts

```diff
@@ -0,0 +1,158 @@
+  INVALID_SPAN_CONTEXT,
+  isValid,
+} from '@opentelemetry/core';
+import { performance } from 'perf_hooks';
+import { TraceOptions } from '@opentelemetry/types';
+
+/**
+ * This class represents a span.
+ */
+export class Span implements types.Span {
+  private _spanContext: types.SpanContext = INVALID_SPAN_CONTEXT;
+  private _tracer: types.Tracer;
+  private _name: string;
+  private _parentId?: string;
+  private _kind: types.SpanKind;
+  private _attributes: types.Attributes = {};
+  private _links: types.Link[] = [];
+  private _events: types.Event[] = [];
+  private _status: types.Status = {
+    code: types.CanonicalCode.OK,
+  };
+  private _ended = false;
+  private _startTime: number;
+  private _duration = 0;
+
+  /** Constructs a new Span instance. */
+  constructor(
+    parentTracer: types.Tracer,
+    spanName: string,
+    options: types.SpanOptions
+  ) {
+    this._tracer = parentTracer;
+    this._name = spanName;
+    this._spanContext.spanId = randomSpanId();
+    this._spanContext.traceOptions = TraceOptions.SAMPLED;
+    const parentSpanContext = this._getParentSpanContext(options.parent);
+    if (parentSpanContext && isValid(parentSpanContext)) {
+      // New child span.
+      this._spanContext.traceId = parentSpanContext.traceId;
+      this._spanContext.traceState = parentSpanContext.traceState;
+      this._parentId = parentSpanContext.spanId;
+    } else {
+      // This is a root span so no remote or local parent.
+      this._spanContext.traceId = randomTraceId();
+    }
+    this._kind = options.kind || types.SpanKind.INTERNAL;
+    this._startTime = options.startTime || performance.now();
+  }
+
+  tracer(): types.Tracer {
+    return this._tracer;
+  }
+
+  context(): types.SpanContext {
+    return this._spanContext;
+  }
+
+  setAttribute(key: string, value: unknown): this {
+    if (this._isSpanEnded()) return this;
+    this._attributes[key] = value;
+    return this;
+  }
+
+  setAttributes(attributes: types.Attributes): this {
+    Object.keys(attributes).forEach(key => {
+      this.setAttribute(key, attributes[key]);
+    });
+    return this;
+  }
+
+  addEvent(name: string, attributes?: types.Attributes): this {
+    if (this._isSpanEnded()) return this;
+    this._events.push({ name, attributes });
+    return this;
+  }
+
+  addLink(spanContext: types.SpanContext, attributes?: types.Attributes): this {
+    if (this._isSpanEnded()) return this;
+    this._links.push({ spanContext, attributes });
+    return this;
+  }
+
+  setStatus(status: types.Status): this {
+    if (this._isSpanEnded()) return this;
+    this._status = status;
+    return this;
+  }
+
+  updateName(name: string): this {
+    if (this._isSpanEnded()) return this;
+    this._name = name;
+    return this;
+  }
+
+  end(endTime?: number): void {
+    if (this._isSpanEnded()) return;
+    this._ended = true;
+    const finishTime = endTime || performance.now();
+    this._duration = finishTime - this._startTime;
```

Comment:
Is storing `duration` rather than `finishTime` something specified by the spec?

Reply:
Good point. This is not really mentioned in the specs, I checked with other OT libraries they are using finishTime instread of duration.

Reply:
done, ptal